src/main/java/picard/sam/SamToFastq.java

```diff
@@ -237,6 +200,13 @@ protected int doWork() {
             writerMapping.closeAll();
         }
 
+        // close all `additionalWriters` only once
+        final Set<FastqWriter> additionalWriterSet = new HashSet<>();
+        additionalWriters.values().forEach(additionalWriterSet::addAll);
+        for (final FastqWriter fastqWriter : additionalWriterSet) {

```

Comment:
It's done this way because you can have duplicate writers in the additional writers collection? Otherwise you could write
```
additionalWriters.values().stream().flatMap(Collection::stream).forEach(FastqWriter::close);
```

Reply:
exactly, if your bam has multiple readgroups but you have `OUTPUT_PER_RG` not set then you will have duplicate writers.